29_Arithmetic2Bench/main.cpp

```diff
@@ -0,0 +1,739 @@
+	BenchmarkSet createBenchmarkPipelines(const smart_refctd_ptr<const ICPUShader>&source, const IGPUPipelineLayout* layout, const uint32_t elementCount, const uint8_t subgroupSizeLog2, const uint32_t workgroupSize, uint32_t itemsPerInvoc = 1u, uint32_t numLoops = 8u)
+	{
+		std::string arith_name = Arithmetic<plus<uint32_t>>::name;
+
+		auto compiler = make_smart_refctd_ptr<asset::CHLSLCompiler>(smart_refctd_ptr(m_system));
+		CHLSLCompiler::SOptions options = {};
+		options.stage = IShader::E_SHADER_STAGE::ESS_COMPUTE;
+		options.targetSpirvVersion = m_device->getPhysicalDevice()->getLimits().spirvVersion;
+		options.spirvOptimizer = nullptr;
+#ifndef _NBL_DEBUG
+		ISPIRVOptimizer::E_OPTIMIZER_PASS optPasses = ISPIRVOptimizer::EOP_STRIP_DEBUG_INFO;
+		auto opt = make_smart_refctd_ptr<ISPIRVOptimizer>(std::span<ISPIRVOptimizer::E_OPTIMIZER_PASS>(&optPasses, 1));
+		options.spirvOptimizer = opt.get();
+#else
+		options.debugInfoFlags |= IShaderCompiler::E_DEBUG_INFO_FLAGS::EDIF_LINE_BIT;
+#endif
+		options.preprocessorOptions.sourceIdentifier = source->getFilepathHint();
+		options.preprocessorOptions.logger = m_logger.get();
+
+		auto* includeFinder = compiler->getDefaultIncludeFinder();
+		includeFinder->addSearchPath("nbl/builtin/hlsl/jit", core::make_smart_refctd_ptr<CJITIncludeLoader>(m_physicalDevice->getLimits(), m_device->getEnabledFeatures()));
+		options.preprocessorOptions.includeFinder = includeFinder;
+
+		const uint32_t subgroupSize = 0x1u << subgroupSizeLog2;
+		const uint32_t itemsPerWG = workgroupSize <= subgroupSize ? workgroupSize * itemsPerInvoc : itemsPerInvoc * max(workgroupSize >> subgroupSizeLog2, subgroupSize) << subgroupSizeLog2;	// TODO use Config somehow
+		smart_refctd_ptr<ICPUShader> overriddenUnspecialized;
+		if constexpr (WorkgroupBench)
+		{
+			const uint32_t workgroupSizeLog2 = hlsl::findMSB(workgroupSize);
+			const std::string definitions[7] = {
+				"workgroup2::" + arith_name,
+				std::to_string(workgroupSizeLog2),
+				std::to_string(itemsPerWG),
+				std::to_string(itemsPerInvoc),
+				std::to_string(subgroupSizeLog2),
+				std::to_string(numLoops),
+				std::to_string(arith_name=="reduction")
+			};
+
+			const IShaderCompiler::SMacroDefinition defines[7] = {
+				{ "OPERATION", definitions[0] },
+				{ "WORKGROUP_SIZE_LOG2", definitions[1] },
+				{ "ITEMS_PER_WG", definitions[2] },
+				{ "ITEMS_PER_INVOCATION", definitions[3] },
+				{ "SUBGROUP_SIZE_LOG2", definitions[4] },
+				{ "NUM_LOOPS", definitions[5] },
+				{ "IS_REDUCTION", definitions[6] }
+			};
+			options.preprocessorOptions.extraDefines = { defines, defines + 7 };
+
+			overriddenUnspecialized = compiler->compileToSPIRV((const char*)source->getContent()->getPointer(), options);
+		}
+		else
+		{
+			const std::string definitions[5] = { 
+				"subgroup2::" + arith_name,
+				std::to_string(workgroupSize),
+				std::to_string(itemsPerInvoc),
+				std::to_string(subgroupSizeLog2),
+				std::to_string(numLoops)
+			};
```

Comment:
also would be awesome if the said `NBL_CONSTEXPR` config had a method to spit out its templated counterpart instantiated type as a std::string

Reply:
can be like
```cpp
template<template<typename> Binop>
using {PERFIX}Config{SUFFIX} = nbl::hlsl::workgroup2::Config<Binop, ... stuff from the runtime struct ....>; 
```

make one for subgroup too